### PR TITLE
Added forcequote command

### LINKED ISSUE #91: add forcequote
For admins that want to force a quote to be quoted as a joke

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: cogs/quote_commands.py [hunk 1]
```diff
@@ -21,19 +21,8 @@ class QuoteCommands(utils.Cog):
...
-        """
-        Quotes a user's message to the guild's quote channel.
-        """
-
-        # Make sure no subcommand is passed
-        if ctx.invoked_subcommand is not None:
-            return
-
+    
+    async def startquote(self, ctx, messages):
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: Change the name of this method - `get_quote_messages` would be more apt.

This method as a whole also needs to be changed entirely anyway - sometimes you're returning a message, sometimes an embed. Be consistent. You can't just copy/paste my code into a new method and hope it works.

File: cogs/quote_commands.py [hunk 2]
```diff
@@ -112,6 +100,24 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
                 embed.description = text
             # embed.set_footer(text=f"Quote ID {quote_id.upper()}")
             embed.timestamp = timestamp
+            return embed
+
+
+        
+
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: What's with all this whitespace my guy? You want one newline seperating functions in classes.

File: cogs/quote_commands.py [hunk 3]
```diff
@@ -112,6 +100,24 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+    @commands.bot_has_permissions(send_messages=True, embed_links=True, add_reactions=True)
+    @commands.guild_only()
+    async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Message]):
+        """
+        Quotes a user's message to the guild's quote channel.
+        """
+              # Make sure no subcommand is passed
+        if ctx.invoked_subcommand is not None:
+            return
+        embed = await startquote(self, ctx, messages)
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: This doesn't account for errored invokes (no messages quoted; invalid messages; wrong authors; etc). This also needs to be `self.startquote`.

File: cogs/quote_commands.py [hunk 4]
```diff
@@ -112,6 +100,24 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+                # Make embed
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: Why is this even still here, and why is it indented?

File: cogs/quote_commands.py [hunk 5]
```diff
@@ -170,6 +176,54 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+            )
+            if rows:
+                return await ctx.send(f"That message has already been quoted with quote ID `{rows[0]['quote_id']}`.")
+            await db(
+                "INSERT INTO user_quotes (quote_id, guild_id, channel_id, message_id, user_id, timestamp) VALUES ($1, $2, $3, $4, $5, $6)",
+                quote_id, ctx.guild.id, posted_message.channel.id, posted_message.id, user.id, timestamp
+            )
+
+        # Output to user
+        await ctx.send(f"{ctx.author.mention}'s quote request saved with ID `{quote_id.upper()}`", embed=embed, ignore_error=True)
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: This code is duplicated and could also be moved out into its own function.

File: cogs/quote_commands.py [hunk 6]
```diff
@@ -170,6 +176,54 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
         # Output to user
         await ctx.send(f"{ctx.author.mention}'s quote request saved with ID `{quote_id.upper()}`", embed=embed, ignore_error=True)
 
+    @utils.command()
+    @commands.has_permissions(manage_messages=True)
+    async def forcequote(self, ctx:utils.Context, messages:commands.Greedy[discord.Message]):
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: Can you change this to a subcommand of `quote` as `quote force`?

**Review by 4Kaylum — CHANGES_REQUESTED** [review]

**Comment by iHeroGH** [comment]
Eyo, what if instead of adding a new `quote force`, add a guild setting like "Allow people with `manage_messages` to bypass the minimum reaction threshold", and then just add a check for that in `quote`?

**Comment by 4Kaylum** [comment]
I'm not sure I want to do that - there might be people out there who selectively like democracy. I dunno. I like them being different.

**Comment by iHeroGH** [comment]
> I'm not sure I want to do that - there might be people out there who selectively like democracy. I dunno. I like them being different.

👌

File: cogs/quote_commands.py [hunk 7]
```diff
@@ -170,6 +188,56 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+            except (discord.Forbidden, AttributeError):
+                pass
+        if quote_channel_id is None or posted_message is None:
+            return await ctx.send("I couldn't send your quote into the quote channel.")
+
+        # And save it to the database
+        async with self.bot.database() as db:
+            rows = await db(
+                "SELECT * FROM user_quotes WHERE guild_id=$1 AND user_id=$2 AND timestamp=$3 AND text=$4",
+                ctx.guild.id, user.id, timestamp, text
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: `user`, `timestamp`, and `text` are undefined.

File: cogs/quote_commands.py [hunk 8]
```diff
@@ -170,6 +188,56 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+            )
+            if rows:
+                return await ctx.send(f"That message has already been quoted with quote ID `{rows[0]['quote_id']}`.")
+            await db(
+                "INSERT INTO user_quotes (quote_id, guild_id, channel_id, message_id, user_id, timestamp) VALUES ($1, $2, $3, $4, $5, $6)",
+                quote_id, ctx.guild.id, posted_message.channel.id, posted_message.id, user.id, timestamp
```
Comment (4Kaylum) [inline cogs/quote_commands.py:211]: `user` and `timestamp` are undefined.

File: cogs/quote_commands.py [hunk 9]
```diff
@@ -170,6 +188,56 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+            )
+
+        # Output to user
+        await ctx.send(f"{ctx.author.mention}'s quote request saved with ID `{quote_id.upper()}`", embed=embed, ignore_error=True)
```
Comment (4Kaylum) [inline cogs/quote_commands.py]: `embed` is undefined.

File: cogs/quote_commands.py [hunk 10]
```diff
@@ -170,6 +188,56 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+            return await ctx.send(response['message'])
+        # Make embed
+        
+
+        # If we get here, we can save to db
+        quote_id = create_id()
+
+        # See if they have a quotes channel
+        quote_channel_id = self.bot.guild_settings[ctx.guild.id].get('quote_channel_id')
+        embed.set_footer(text=f"Quote ID {quote_id.upper()}")
```
Comment (4Kaylum) [inline cogs/quote_commands.py:196]: `embed` is undefined.

File: cogs/quote_commands.py [hunk 11]
```diff
@@ -170,6 +188,56 @@ async def quote(self, ctx:utils.Context, messages:commands.Greedy[discord.Messag
...
+        posted_message = None
+        if quote_channel_id:
+            channel = self.bot.get_channel(quote_channel_id)
+            try:
+                posted_message = await channel.send(embed=embed)
```
Comment (4Kaylum) [inline cogs/quote_commands.py:201]: `embed` is undefined.

**Review by 4Kaylum — CHANGES_REQUESTED** [review]
You've fixed some of the issues I had with your code, but now there are actual _errors_ in the code. A lot of them are invalid variables for issues that you fixed in the _normal_ quote command, but not the `quote force` command. 

Generally: embed isn't defined, user isn't defined, timestamp isn't defined.

**Pull request merged**